haystack/nodes/base.py

```diff
@@ -1,36 +1,81 @@
 from __future__ import annotations
-from typing import Any, Callable, Optional, Dict, List, Tuple, Optional
+from typing import Any, Optional, Dict, List, Tuple, Optional
 
-import io
-from functools import wraps
+import sys
 from copy import deepcopy
-from abc import abstractmethod
+from abc import ABC, ABCMeta, abstractmethod
+from functools import wraps
 import inspect
 import logging
 
 from haystack.schema import Document, MultiLabel
+from haystack.errors import PipelineError
 
 
 logger = logging.getLogger(__name__)
 
 
-class BaseComponent:
+def exportable_to_yaml(init_func):
+    """
+    Decorator that saves the init parameters of a node that later can
+    be used with exporting YAML configuration of a Pipeline.
+    """
+
+    @wraps(init_func)
+    def wrapper_exportable_to_yaml(self, *args, **kwargs):
+
+        # Call the actuall __init__ function with all the arguments
+        init_func(self, *args, **kwargs)
+
+        # Warn for unnamed input params - should be rare
+        if args:
+            logger.warning(
+                "Unnamed __init__ parameters will not be saved to YAML if Pipeline.save_to_yaml() is called!"
+            )
+
+        # with super().__init__() this method might be executed many times
+        # FIXME it would be better to decorate only non-abstract classes though!
+        if not self._pipeline_config:
+            self._pipeline_config = {"params": {}, "type": type(self).__name__}
+
+        # Store all the named input parameters in self._pipeline_config
+        for k, v in kwargs.items():
+            if isinstance(v, BaseComponent):
+                self._pipeline_config["params"][k] = v._pipeline_config
+            elif v is not None:
+                self._pipeline_config["params"][k] = v
+
+    return wrapper_exportable_to_yaml
+
+
+# Metaclasses are like decorators for classes
+# __new__ is called when a class is created (not an instance!)
+# Inherits from ABCMeta to avoid metaclass conflicts with ABC
+class Meta(ABCMeta):
+    def __new__(cls, name, bases, dct):
+        subclass = super().__new__(cls, name, bases, dct)
+
+        # Automatically registers all the init parameters in
+        # an instance attribute called `_pipeline_config`,
+        # used tosave this component to YAML. See exportable_to_yaml()
+        subclass.__init__ = exportable_to_yaml(subclass.__init__)
+
+        # Keeps track of all available subclasses by name.
+        # Enables generic load() for all specific component implementations.
+        subclass.subclasses[subclass.__name__] = subclass
```

Comment:
Unfortunately `__init_subclass__` does not seem to get the init arguments (or at least I could find a way to access them) so it could not be used for `exportable to yaml`. In addition I moved the subclasses line in here because somehow ABC modified the behavior of `__init_subclass__` just enough to break it. I don't remember exactly what broke though, so I will try it out again to see if it can work as it was before.

Another thing to note is that, unless `subclasses` is explicitly set to `None` in a custom node, or deleted explicitly in a competing metaclass, it's always defined, because it's a class attribute. It's unlikely that custom nodes will delete it: we can even make it private (`_subclasses`) to make sure, but I wouldn't worry about it not being defined.

In general I could rather improve the metaclass to work only on non-abstract classes with something like:
```python
class Meta(ABCMeta):
    def __new__(cls, name, bases, dct):
        subclass = super().__new__(cls, name, bases, dct)

        # If the class does not inherit from ABC
        if not isinstance(cls.__class__, ABCMeta):    # TO BE TESTED

                # Automatically registers all the init parameters in
                # an instance attribute called `_pipeline_config`,
                # used tosave this component to YAML. See exportable_to_yaml()
                subclass.__init__ = exportable_to_yaml(subclass.__init__)
        
                # Keeps track of all available subclasses by name.
                # Enables generic load() for all specific component implementations.
                subclass.subclasses[subclass.__name__] = subclass

        return subclass
```
I'll have to see if this is viable and solves the issue though.

Another pro of the metaclass approach is that this all thing won't run at execution time like `__init_subclass__`, but only once at class import time. So I'd consider it a minuscule performance improvement (even though negligible, I''l admit).

Reply:
Yes, __init_subclass__ does get all you need. And it is also called during ABCMeta's __new__. Look at this stacktrace:
```
__init_subclass__ (/home/tstad/git/haystack/haystack/nodes/base.py:84)
__new__ (/home/tstad/miniconda3/envs/haystack-dev2/lib/python3.7/abc.py:126)
<module> (/home/tstad/git/haystack/haystack/nodes/file_converter/base.py:10)
<module> (/home/tstad/git/haystack/haystack/nodes/file_converter/__init__.py:1)
<module> (/home/tstad/git/haystack/haystack/utils/preprocessing.py:7)
<module> (/home/tstad/git/haystack/haystack/utils/__init__.py:1)
<module> (/home/tstad/git/haystack/haystack/nodes/__init__.py:1)
<module> (/home/tstad/git/haystack/haystack/pipelines/base.py:23)
<module> (/home/tstad/git/haystack/haystack/pipelines/__init__.py:1)
<module> (/home/tstad/git/haystack/haystack/__init__.py:21)
<module> (/home/tstad/git/haystack/debug_meta.py:1)
_run_code (/home/tstad/miniconda3/envs/haystack-dev2/lib/python3.7/runpy.py:85)
_run_module_code (/home/tstad/miniconda3/envs/haystack-dev2/lib/python3.7/runpy.py:96)
run_path (/home/tstad/miniconda3/envs/haystack-dev2/lib/python3.7/runpy.py:263)
_run_code (/home/tstad/miniconda3/envs/haystack-dev2/lib/python3.7/runpy.py:85)
_run_module_as_main (/home/tstad/miniconda3/envs/haystack-dev2/lib/python3.7/runpy.py:193)
```

So basically you can achieve everything by simply having `BaseComponent` inheriting from `ABC` and implementing __init_subclass__ like that:
```python
    def __init_subclass__(cls, **kwargs):
        """
        Automatically keeps track of all available subclasses.
        Enables generic load() for all specific component implementations.
        """
        super().__init_subclass__(**kwargs)
        cls.__init__ = exportable_to_yaml(cls.__init__)
        cls.subclasses[cls.__name__] = cls
```
There's no need for an additional Metaclass. 
Btw, I'm not sure if I wasn't clear enough about that. But the `Meta` class currently only works with classes that have a `subclass` field, which makes it too tightly coupled to `BaseComponent` IMHO. Checkout:
```python
class TestComponent(metaclass=Meta):
    pass
```
importing this class will fail during `class_.subclasses[class_.__name__] = class_` of `Meta` as it has no `subclasses` field.

Reply:
Ooh, great! I get it now. I definitely did not understand when `__init_subclass__` was called :smile: I'll move the implementation to `__init_subclass__` then :+1: